testscript/exe.go

```diff
@@ -88,7 +89,11 @@ func Main(m TestingM, commands map[string]func()) {
 				ts.cmdExec(neg, append([]string{name}, args...))
 			}
 		}
-		os.Exit(m.Run())
+		exit := m.Run()
+		// The [os.Exit] below prevents defers from running, so we must clean up here.
+		// The defer above is still useful in case we fail before we get to this point.
+		cleanup()
```

Comment:
Alternatively (and arguably slightly cleaner) you could factor all the `Main` code into a separate function (e.g. `main1`) that returns int, and change `Main` to just do `os.Exit(main1(...))`

Reply:
done; PTAL again as I ended up only factoring out part of the `Main` code, given that the other branch has no control over whether the user's command calls `os.Exit`.